core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1517,9 +1517,12 @@ class SparkContext(config: SparkConf) extends Logging {
    * only supported for Hadoop-supported filesystems.
    */
   def addFile(path: String, recursive: Boolean): Unit = {
-    val uri = new Path(path).toUri
+    var uri = new Path(path).toUri
     val schemeCorrectedPath = uri.getScheme match {
-      case null | "local" => new File(path).getCanonicalFile.toURI.toString
+      case null | "local" =>
+        // SPARK-24195: Local is not a valid scheme for FileSystem, we should only keep path here.
+        uri = new Path(uri.getPath).toUri
```

Comment:
Why is this needed? Can't we just do `new File(uri.getPath).getCanonicalFile.toURI.toString` without this line?

Reply:
Yes, same question. The above line seems not useful.

Reply:
it changes `uri` - which is reference again below.

Reply:
Yes, just as @felixcheung said, this because we will use uri in https://github.com/apache/spark/pull/21533/files/f922fd8c995164cada4a8b72e92c369a827def16#diff-364713d7776956cb8b0a771e9b62f82dR1557, if the uri with local scheme, we'll get an exception cause local is not a valid scheme for FileSystem.

Reply:
I mean we `getPath` doesn't include scheme:

```
scala> new Path("local://a/b/c")
res0: org.apache.hadoop.fs.Path = local://a/b/c

scala> new Path("local://a/b/c").toUri
res1: java.net.URI = local://a/b/c

scala> new Path("local://a/b/c").toUri.getScheme
res2: String = local

scala> new Path("local://a/b/c").toUri.getPath
res3: String = /b/c
```

why should we do this again?

```
scala> new Path(new Path("local://a/b/c").toUri.getPath).toUri.getPath
res4: String = /b/c
```

Reply:
Yea we can simplify this.

Reply:
@HyukjinKwon @jiangxb1987 
Thanks for your explain, I think I know what's your meaning about `we getPath doesn't include scheme`. Actually the purpose of this code `uri = new Path(uri.getPath).toUri`, is to reassign the var in +1520, we don't want the uri including local scheme.
```
Can't we just do new File(uri.getPath).getCanonicalFile.toURI.toString without this line?
```
We can't because like I explained above, if we didn't do `uri = new Path(uri.getPath).toUri`, will get a exception like below:
```
No FileSystem for scheme: local
java.io.IOException: No FileSystem for scheme: local
	at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2586)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2593)
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:91)
	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2632)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2614)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:370)
	at org.apache.spark.util.Utils$.getHadoopFileSystem(Utils.scala:1830)
	at org.apache.spark.util.Utils$.doFetchFile(Utils.scala:690)
	at org.apache.spark.util.Utils$.fetchFile(Utils.scala:486)
	at org.apache.spark.SparkContext.addFile(SparkContext.scala:1557)
```

Reply:
I mean, at least we can do:

```
val a = new File(uri.getPath).getCanonicalFile.toURI.toString
uri = new Path(uri.getPath).toUri
a
```

`new Path(uri.getPath).toUri` for trimming the scheme looks not quite clean though. It's a-okay at least to me.

Reply:
Ah, I see, thanks. I'll do this in the next commit. Thanks for your patient explain.